bleak/backends/dotnet/client.py

```diff
@@ -168,79 +170,117 @@ async def connect(self, **kwargs) -> bool:
             return_type=BluetoothLEDevice,
         )
 
+        # Called on disconnect event or on failure to connect.
+        def handle_disconnect():
+            if self._connection_status_changed_token:
+                self._requester.remove_ConnectionStatusChanged(
+                    self._connection_status_changed_token
+                )
+                self._connection_status_changed_token = None
+
+            if self._requester:
+                self._requester.Dispose()
+                self._requester = None
+
+            if self._session:
+                self._session.Dispose()
+                self._session = None
+
+        def handle_connection_status_changed(
+            connection_status: BluetoothConnectionStatus,
+        ):
+            if connection_status == BluetoothConnectionStatus.Connected:
+                for e in self._connect_events:
+                    e.set()
+
+            elif connection_status == BluetoothConnectionStatus.Disconnected:
+                if self._disconnected_callback:
+                    self._disconnected_callback(self)
+
+                for e in self._disconnect_events:
+                    e.set()
+
+                handle_disconnect()
+
         loop = asyncio.get_event_loop()
 
         def _ConnectionStatusChanged_Handler(sender, args):
             logger.debug(
                 "_ConnectionStatusChanged_Handler: %d", sender.ConnectionStatus
             )
-            if (
-                sender.ConnectionStatus == BluetoothConnectionStatus.Disconnected
-                and self._disconnected_callback
-            ):
-                loop.call_soon_threadsafe(self._disconnected_callback, self)
-
-        self._requester.ConnectionStatusChanged += _ConnectionStatusChanged_Handler
+            loop.call_soon_threadsafe(
+                handle_connection_status_changed, sender.ConnectionStatus
+            )
 
-        # Obtain services, which also leads to connection being established.
-        services = await self.get_services()
```

Comment:
If `await self.get_services()` is not called in `connect`, it is required of the user to call it before trying to do any other operations. It should be done here to make sure that the `BleakClient` is ready for anything after the connection.

Reply:
Fixed. Thanks.